zkml/src/parser/json.rs

```diff
@@ -0,0 +1,329 @@
+        let embedding_size = hidden_size;
+        let num_heads = l.metadata_to_u32("num_attention_heads")? as usize;
+        let num_blocks = l.metadata_to_u32("num_hidden_layers")? as usize;
+        let context_length = l.metadata_to_u32("max_seq_len")? as usize;
+        let norm_epsilon = l.metadata_to_f32("norm_epsilon")?;
+        Ok(Self {
+            embedding_size,
+            hidden_size,
+            num_heads,
+            num_block: num_blocks,
+            context_length,
+            norm_epsilon,
+            specific_config: variant,
+        })
+    }
+}
+
+impl LLMVariant {
+    pub fn from_json(l: &FileTensorLoader) -> anyhow::Result<Self> {
+        let variant_value = l
+            .get_metadata("model_name")
+            .ok_or_else(|| anyhow::anyhow!("Metadata key 'model_name' not found"))?;
+
+        let model_name_str = variant_value
+            .as_str()
+            .ok_or_else(|| anyhow::anyhow!("Metadata 'model_name' is not a string value"))?;
+
+        match model_name_str
+            .trim() // Keep the trim, it's good practice
+        {
+            "gpt2" => Ok(Self::GPT2),
+            "sshleifer/tiny-gpt2" => Ok(Self::GPT2),
+            a => bail!("unsupported architecture: {:?}", a),
+        }
+    }
+}
+
+impl FeedForward<f32> {
+    pub fn from_json(l: &FileTensorLoader, c: &LLMConfig) -> anyhow::Result<Self> {
+        let norm = LayerNorm::from_json(&l.pp("ffn_"), c)?;
+        let up = l.get_tensor("ffn_up.weight")?;
+        let up_bias = l.get_tensor("ffn_up.bias")?;
+        let down = l.get_tensor("ffn_down.weight")?;
+        let down_bias = l.get_tensor("ffn_down.bias")?;
+        ensure!(
+            down.get_shape()[0] == c.embedding_size,
+            "down must have shape {:?} vs embedding_size: {}",
+            down.get_shape(),
+            c.embedding_size
+        );
+        Ok(Self {
+            norm,
+            up,
+            up_bias,
+            down,
+            down_bias,
+        })
+    }
+}
+
+impl Attention<f32> {
+    pub fn from_json(l: &FileTensorLoader, c: &LLMConfig) -> anyhow::Result<Self> {
+        let norm = LayerNorm::from_json(&l.pp("attn_"), c)
+            .context("Failed to load LayerNorm for attention in from_json")?;
+
+        let fused_qkv_weight = l
+            .get_tensor("attn_qkv.weight")
+            .context("Failed to load attn_qkv.weight in from_json")?;
+        let fused_qkv_bias = l
+            .get_tensor("attn_qkv.bias")
+            .context("Failed to load attn_qkv.bias in from_json")?;
+
+        let hidden_size = c.hidden_size; // embedding_dim for GPT-2
+
+        // Unfuse weights:
+        // Expected shape of fused_qkv_weight is [3 * hidden_size, hidden_size] after python script transpose.
+        // Each individual q, k, v weight matrix should be [hidden_size, hidden_size].
+        // So, each chunk has hidden_size * hidden_size elements.
+        let weight_chunk_elements = hidden_size * hidden_size;
```

Comment:
Shouldn't this be 'c.embedding_size * hidden_size`? The unfused tensors later are assumed to have this shape